### ISSUE TITLE
StreamingPull various errors and exceptions, requesting advice on how to handle

### ISSUE THREAD

**Issue by hach-que** [GoogleCloudPlatform/google-cloud-dotnet#1653]
I'm not sure if the logic we currently have around using StreamingPull is correct for a background service. We regularly see "InvalidOperationException: Shutdown has already been called", "ObjectDisposedException: Safe handle has been closed" and gRPC error code Aborted in our use of the StreamingPull function. I've based our code roughly off the provided example, but the provided example doesn't actually loop so it's hard to tell what the right logic should be.

I've attached the code below we currently have for continuous polling over multiple topics. You can see the bunch of exceptions we're currently handling and dropping back to the Pull() fallback (which just does an immediate check for pending messages via the non-streaming method and then goes back to trying streaming).

My concern is that exceptions like "InvalidOperationException: Shutdown has already been called" mean the gRPC connection itself is broken and that we're supposed to be completely recreating the client. Or that by ignoring these exceptions we're going to run into situations where our background services will silently stop processing messages because we aren't handling them properly.

We even see "Shutdown has already been called" on the WriteAsync used to ack messages inside the "await onMessage" block. This call is indicates the connection was closed sometime between getting the message through streaming pull and ack'ing it, but our code never explicitly closes the connection (we expect it to stay open pretty much constantly, and certainly not in the small time between pulling a message an ack'ing it). Are we expected to fall back to the non-streaming ack method here if WriteAsync fails?

If someone could provide advice as to the correctness of this polling code, and what we should be doing to handle these exceptions / gRPC error codes for streaming pulls, that would be greatly appreciated.

```csharp
public async Task SubscribeAndLoopForever(string subscriptionName, string[] eventTypes, Func<SerializedEvent, Task<bool>> onMessage)
{
    // Subscribe to all of the topics for all of the event types in parallel.
    var tasks = new Task[eventTypes.Length];
    for (var i = 0; i < eventTypes.Length; i++)
    {
        var eventType = eventTypes[i];

        var topicName = new TopicName(
            HiveMPEnv.GetGoogleProjectId(),
            "event~" + eventType.Replace(":", "."));
        var subscriberName = new SubscriptionName(
            HiveMPEnv.GetGoogleProjectId(),
            subscriptionName + "~" + eventType.Replace(":", "."));

        try
        {
            await _subscriberClient.CreateSubscriptionAsync(
                subscriberName,
                topicName,
                null,
                600);
        }
        catch (RpcException ex) when (ex.Status.StatusCode == StatusCode.NotFound)
        {
            try
            {
                // The topic wasn't found, so create it.
                await _publisherClient.CreateTopicAsync(topicName);
            }
            catch (RpcException ex2) when (ex2.Status.StatusCode == StatusCode.AlreadyExists)
            {
                // Topic has been created in parallel.
            }

            try
            {
                // Now create the subscription.
                await _subscriberClient.CreateSubscriptionAsync(
                    subscriberName,
                    topicName,
                    null,
                    600);
            }
            catch (RpcException ex2) when (ex2.Status.StatusCode == StatusCode.AlreadyExists)
            {
                // Subscription already exists; everything is OK.
            }
        }
        catch (RpcException ex) when (ex.Status.StatusCode == StatusCode.AlreadyExists)
        {
            // Subscription already exists; everything is OK.
        }

        // Set up the task to continously poll.
        tasks[i] = Task.Run(async () =>
        {
            _logger.LogInformation($"Beginning poll for " + eventType + " events from " + subscriptionName + "...");
            while (true)
            {
                var useFallback = false;
                try
                {
                    var stream = _subscriberClient.StreamingPull();
                    try
                    {
                        await stream.WriteAsync(new StreamingPullRequest
                        {
                            SubscriptionAsSubscriptionName = subscriberName,
                            StreamAckDeadlineSeconds = 600,
                        });
                        while (await stream.ResponseStream.MoveNext())
                        {
                            var messages = stream.ResponseStream.Current;
                                    
                            foreach (var message in messages.ReceivedMessages)
                            {
                                var didAck = false;
                                try
                                {
                                    var serializedEvent = JsonConvert.DeserializeObject<SerializedEvent>(
                                        message.Message.Data.ToStringUtf8());
                                    if (await onMessage(serializedEvent))
                                    {
var useAckFallback = false;
                                        try
                                        {
                                            var req = new StreamingPullRequest();
                                            req.AckIds.Add(message.AckId);
                                            await stream.WriteAsync(req);
                                        }
                                        catch (RpcException sx) when (sx.Status.StatusCode == StatusCode.Unimplemented || sx.Status.StatusCode == StatusCode.Unavailable || sx.Status.StatusCode == StatusCode.Aborted)
                                        {
                                            useAckFallback = true;
                                        }
                                        catch (InvalidOperationException ex) when (ex.Message.Contains("Shutdown has already been called"))
                                        {
                                            useAckFallback = true;
                                        }
                                        catch (ObjectDisposedException ex) when (ex.Message.Contains("Safe handle has been closed"))
                                        {
                                            useAckFallback = true;
                                        }

                                        if (useAckFallback)
                                        {
                                            // Stream died while trying to ack through streaming channel; fallback
                                            // to manual request to ack message so we don't forget to ack it because
                                            // we have processed it at this point.
                                            var req = new AcknowledgeRequest();
                                            req.SubscriptionAsSubscriptionName = subscriberName;
                                            req.AckIds.Add(message.AckId);
                                            await _subscriberClient.AcknowledgeAsync(req);
                                        }

                                        didAck = true;
                                    }
                                }
                                catch (Exception ex)
                                {
                                    _errorReporting.Report(ex);
                                }
                                finally
                                {
                                    if (!didAck)
                                    {
                                        // We didn't process the message; make it immediately available to other consumers.
                                        var req = new StreamingPullRequest();
                                        req.ModifyDeadlineAckIds.Add(message.AckId);
                                        req.ModifyDeadlineSeconds.Add(0);
                                        await stream.WriteAsync(req);
                                    }
                                }
                            }
                        }
                    }
                    finally
                    {
                        await stream.WriteCompleteAsync();
                    }
                }
                catch (RpcException sx) when (sx.Status.StatusCode == StatusCode.Unimplemented || sx.Status.StatusCode == StatusCode.Unavailable || sx.Status.StatusCode == StatusCode.Aborted)
                {
                    useFallback = true;
                }
                catch (InvalidOperationException ex) when (ex.Message.Contains("Shutdown has already been called"))
                {
                    useFallback = true;
                }
                catch (ObjectDisposedException ex) when (ex.Message.Contains("Safe handle has been closed"))
                {
                    useFallback = true;
                }
                catch (Exception ex)
                {
                    _errorReporting.Report(ex);
                }

                if (useFallback)
                {
                    // Streaming pull not supported, fallback to pull operation.
                    if (HiveMPEnv.GetShouldUseProductionMetricForPubSubFailures())
                    {
                        await _metricService.AddPoint<ProdGooglePubSubStreamingFailuresMetric>();
                    }
                    else
                    {
                        await _metricService.AddPoint<DevGooglePubSubStreamingFailuresMetric>();
                    }
                    try
                    {
                        var messages = await _subscriberClient.PullAsync(new PullRequest
                        {
                            MaxMessages = 60,
                            ReturnImmediately = true,
                            SubscriptionAsSubscriptionName = subscriberName,
                        });

                        var toAck = new List<string>(messages.ReceivedMessages.Count);
                        foreach (var message in messages.ReceivedMessages)
                        {
                            try
                            {
                                var serializedEvent = JsonConvert.DeserializeObject<SerializedEvent>(
                                    message.Message.Data.ToStringUtf8());
                                if (await onMessage(serializedEvent))
                                {
                                    toAck.Add(message.AckId);
                                }
                            }
                            catch (Exception ex)
                            {
                                _errorReporting.Report(ex);
                            }
                        }

                        if (toAck.Count > 0)
                        {
                            var req = new AcknowledgeRequest();
                            req.SubscriptionAsSubscriptionName = subscriberName;
                            req.AckIds.AddRange(toAck);
                            await _subscriberClient.AcknowledgeAsync(req);
                        }
                    }
                    catch (InvalidOperationException ex) when (ex.Message.Contains("Shutdown has already been called"))
                    {
                        // We'll retry in a second.
                    }
                    catch (ObjectDisposedException ex) when (ex.Message.Contains("Safe handle has been closed"))
                    {
                        // We'll retry in a second.
                    }
                    catch (Exception ex)
                    {
                        _errorReporting.Report(ex);
                    }
                }

                await Task.Delay(1000);
            }
        });
    }

    await Task.WhenAll(tasks);
}
```

**Comment by manigandham** [comment]
We also saw this same issue...

**Comment by chrisdunelm** [comment]
@hach-que @manigandham thanks for reporting these issues.

To help me reproduce this, please could you let me know:

* What environment this is running in - e.g. Windows console app, .NET Core console app, ASP.NET, ASP.NET Core, ...
* Roughly how many messages per minute are you receiving? Are you messages "bursty"? I.e. Do you receive a lot of messages, then long gaps; or are messages received are a fairly continuous rate?
* Do unexpected exceptions occur at random times? Or after the subscriber has been active for a certain time? Or after a gap in receiving messages?

This will help me get an idea of your messaging characteristics.

Have you tried using `SimpleSubscriber`? It's intended

**Comment by manigandham** [comment]
Console apps using asp.net framework on .net core running in containers on top of `microsoft\aspnetcore` image. Listening to ctrl+c or other sigterm signal to shutdown. Around 5k msgs/sec constant average.

Errors for us were on shutdown where it would fail with this error no matter how safely we tried to cancel the consumer (using timeouts, locks, single-thread, etc). We used SimpleSubscriber, which as far as I know uses streaming pull underneath. Didn't use streaming pull directly. When doing normal pull/ack using subscriber client, we didnt have any errors. We stopped using pub/sub so don't have recent data for you unfortunately.

**Comment by chrisdunelm** [comment]
@manigandham thanks for the details.
Just to clarify, did you *only* have problems on `SimpleSubscriber` shutdown? So messages were received and acked OK, but you had exceptions on shutdown?

Yes, you are correct that `SimpleSubscriber` uses streaming pull.

**Comment by manigandham** [comment]
From what I remember, yes it was during the shutdown. We did see message redelivery, unknown if that was only from shutdown error or not.

**Comment by chrisdunelm** [comment]
@hach-que It is expected that a `StreamingPull()` connection will periodically be closed by the server. The client code must expect this, and re-start the streaming-pull if the error is recoverable. See [Extensions. IsRecoverable()](https://github.com/GoogleCloudPlatform/google-cloud-dotnet/blob/master/apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1/Extensions.cs#L68) for a list of recoverable errors.

Once the server has shut the connection, it is expected that calls to `WriteAsync()` and `WriteCompleteAsync()` will throw; as the connection is already shut down.

It should never be necessary to fallback to polling.

I would very strongly suggest that you use the [`SimpleSubscriber`](https://github.com/GoogleCloudPlatform/google-cloud-dotnet/blob/master/apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1/SimpleSubscriber.cs) class, as it is designed to many of these details. It handles ack'ing and lease-extending messages; re-starting `streamingPull()` calls; and supports high-performance by managing multiple subscription channels concurrently.

**Comment by chrisdunelm** [comment]
Closing. Please re-open if there is still a problem.

**Comment by hach-que** [comment]
@chrisdunelm I'm currently trying to move over to SimpleSubscriber, however with the Pub/Sub emulator it never seems to fire handler. Looking at the code, it doesn't have any fallback to PullAsync if the Pub/Sub endpoint reports NotSupported, which I believe is the case for the emulator.

Can you confirm that SimpleSubscriber works when using the Cloud Pub/Sub Emulator? If it doesn't, then SimpleSubscriber not a viable solution for us.

**Comment by chrisdunelm** [comment]
@hach-que Apologies for the late reply. We were all on vacation over Christmas and New Year.
Just to check, you're talking about the emulator here: https://cloud.google.com/pubsub/docs/emulator ?
You are correct that the `SimpleSubscriber` does not fallback to `PullAsync` if the streaming call fails. This code has not been tested against the emulator, only the real cloud service.
I've created #1799 to track work to support the emulator.

**Comment by hach-que** [comment]
@chrisdunelm Hi, I ended up getting it working against the emulator a couple of hours ago. It turns out I forgot to explicitly pass the credentials and service endpoint to the SimpleSubscriber (which is required when you want to point it at the emulator). Once I did that, it started working again. I was going to post an update here once it went out to production, but that hasn't happened yet.

For anyone else that wants to point it at the emulator, something like this is required:

```
                    var simpleSubscriber = await SimpleSubscriber.CreateAsync(subscriberName, new SimpleSubscriber.ClientCreationSettings(
                        clientCount: 1,
                        subscriberSettings: null,
                        credentials: ChannelCredentials.Insecure,
                        serviceEndpoint: new ServiceEndpoint(emulatorHost, emulatorPort)));
```

**Comment by chrisdunelm** [comment]
@hach-que Great, glad to hear it's working, and thanks for the code example. I'll close #1799, as it sounds like there's nothing to do :)
